runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -372,52 +393,55 @@ function STHighlighter:on_win(topline, botline)
+          hl_group = hl_group,
+          end_col = token.end_col,
+          priority = vim.highlight.priorities.semantic_tokens + delta,
+          strict = false,
+        })
+      end
+
+      local ft = vim.bo[self.bufnr].filetype
       local highlights = current_result.highlights
-      local idx = binary_search(highlights, topline)
+      local first = lower_bound(highlights, topline, 1, #highlights + 1)
+      local last = upper_bound(highlights, botline, first, #highlights + 1) - 1
 
-      for i = idx, #highlights do
+      for i = first, last do
         local token = highlights[i]
-
-        if token.line > botline then
-          break
-        end
-
-        if not token.extmark_added then
-          -- `strict = false` is necessary here for the 1% of cases where the
-          -- current result doesn't actually match the buffer contents. Some
-          -- LSP servers can respond with stale tokens on requests if they are
-          -- still processing changes from a didChange notification.
-          --
-          -- LSP servers that do this _should_ follow up known stale responses
-          -- with a refresh notification once they've finished processing the
-          -- didChange notification, which would re-synchronize the tokens from
-          -- our end.
-          --
-          -- The server I know of that does this is clangd when the preamble of
-          -- a file changes and the token request is processed with a stale
-          -- preamble while the new one is still being built. Once the preamble
-          -- finishes, clangd sends a refresh request which lets the client
-          -- re-synchronize the tokens.
-          api.nvim_buf_set_extmark(self.bufnr, state.namespace, token.line, token.start_col, {
-            hl_group = '@' .. token.type,
-            end_col = token.end_col,
-            priority = vim.highlight.priorities.semantic_tokens,
-            strict = false,
-          })
-
-          -- TODO(bfredl) use single extmark when hl_group supports table
-          if #token.modifiers > 0 then
-            for _, modifier in pairs(token.modifiers) do
-              api.nvim_buf_set_extmark(self.bufnr, state.namespace, token.line, token.start_col, {
-                hl_group = '@' .. modifier,
-                end_col = token.end_col,
-                priority = vim.highlight.priorities.semantic_tokens + 1,
-                strict = false,
-              })
-            end
+        if not token.marked then
+          set_mark(token, string.format('@lsp.type.%s.%s', token.type, ft), 0)
+          for modifier, _ in pairs(token.modifiers) do
+            set_mark(token, string.format('@lsp.mod.%s.%s', modifier, ft), 1)
+            set_mark(token, string.format('@lsp.both.%s.%s.%s', token.type, modifier, ft), 2)
           end
-
-          token.extmark_added = true
+          token.marked = true
+
+          api.nvim_exec_autocmds('LspTokenUpdate', {
```

Comment:
perhaps this has been discussed before, but I think the behavior would be predictable if 

1. this code was moved before the `@lsp.foo.bar` code
2. `highlight_token()` would set `token.marked` (or at least provide an option to do so)
3. after the autocmd,  we check again if `token.marked` was set and then _disable_ the setting of the default highlight groups ( i e `@lsp.whatever` ) if this was done.

This would make it easier to completely override the behavior of a token without needing to counteract the effect of the default groups (which might set colors and attrs not set by the autocmd provided hl_group).

Reply:
Bikeshed, while we're at it: not crazy about `both` -- maybe `typemod` or `combined` or `combination` or somesuch would be more descriptive.

Reply:
> perhaps this has been discussed before, but I think the behavior would be predictable if 
> 
> 
> 
> 1. this code was moved before the `@lsp.foo.bar` code
> 
> 2. `highlight_token()` would set `token.marked` (or at least provide an option to do so)
> 
> 3. after the autocmd,  we check again if `token.marked` was set and then _disable_ the setting of the default highlight groups ( i e `@lsp.whatever` ) if this was done.
> 
> 
> 
> This would make it easier to completely override the behavior of a token without needing to counteract the effect of the default groups (which might set colors and attrs not set by the autocmd provided hl_group).

I really like this idea, and might even consider it a must. The one part of this PR I'm not super excited about is that the default extmarks are being added no matter what a user has in their autocmd. This would resolve that nicely.

Reply:
Wouldn't this also mean that as soon as you want to have a single customisation you'd also need to apply all defaults as well if color schemes start using these? 

I'd think it's more confusing if other highlights disappear. Afaik this was also discussed before. If there are no new substantial new arguments for a behaviour change I'd say we go ahead with this and let people actually try and use it before having another round of discussion going back and forth

---

> Bikeshed, while we're at it: not crazy about both -- maybe typemod or combined or combination or somesuch would be more descriptive.

+1 for typemod

Reply:
> Wouldn't this also mean that as soon as you want to have a single customisation you'd also need to apply all defaults as well if color schemes start using these? 


Not if marking the token as added in the utility highlight function was optional (and false by default). Just giving the ability to override defaults for a particular token would be extremely useful and wouldn't really change the behavior this PR currently provides.

Reply:
> Not if marking the token as added in the utility highlight function was optional (and false by default)

Right, this seems like the correct resolution. The default behavior should remain like it is, but we provide a realiable opt-in _mechanism_ to completely override highlights for user who understand what it implies and wants that.

Reply:
I agree that this would be useful and straightforward to implement, but I am fine with leaving it for a future PR.

I imagine that the `opts` argument would add a second field; something like:
``` lua
---@param opts (table|nil) Optional parameters.
---       - priority: (number|nil) Priority for the applied extmark. Defaults
---         to `vim.highlight.priorities.semantic_tokens + 3`
---       - keep_default (boolean|nil) Defaults to true. If true, apply the
---         default `@lsp.*` highlights as usual. If false, do not apply the
---         default highlights for this token.
```

I don't think this is the same as previous discussion about "disabling the default rule."  That was about disabling all default semantic highlighting, rather than allowing you to disable it per-token using the callback.

There was also speculation about a hypothetical change enabling return values from autocommand callbacks, and how that would obviate the need for a separate `highlight_token()` function. Adding more logic to `highlight_token()` doesn't make that impossible, but maybe less clean.

Reply:
> I don't think this is the same as previous discussion about "disabling the default rule." That was about disabling all default semantic highlighting, rather than allowing you to disable it per-token using the callback.

Oh right. 

I don't have a strong opinion on this but I'd really like to defer changing this to a separate PR. 
If there are no other objections I'll go ahead and merge this in a bit